client/components/header-button/style.scss

```diff
@@ -1,4 +1,4 @@
-.header-button {
+.button.header-button {
```

Comment:
Unortunately, I think this goes a bit against our selector principles -- I think for a component called `HeaderButton` we want `.header-button` to be specific enough.

There's also a fair chance that a number of other components 'inherit' styling from other ones, that would require the same kind of change, i.e. `.underlying-component.current-component`, which could become pretty messy.

I'd like to get some input from a designer who's acquainted with our CSS selector rules, and if there's any other way around this problem.

Long time no cc @folletto :grin: (see PR desc -- right now, we rely on the order of `Button` styling to be imported before `HeaderButton`, which we can no longer guarantee with a webpack loader based style build, which is what we want/need to implement -- for the Calypso/Gutenberg SDK, among others)

Reply:
Yes I share the same view. Each item should be top level, specific, and without superfluous combinations... so this change while might be working now, it has potential to break the maintenance later.

I feel the issue is here: `<Button className="header-button" { ...rest }>` as this adds a class to `Button` instead of wrapping it `<div className="header-button"><Button ...></div>`.
I'm not sure either if wrapping is the _ideal_ solution here, but it might solve the problem as the specificity is fixed by cascade (`.header-button .button` vs `.button.header-button`). 

That said... there's also a design pattern / architecture consideration:
1. Button should be self-contained and define only the patterns that are available, and should never be overridden
2. Button has an "unstyled" version `ButtonCore`, and then `Button` itself uses it adding style.

This consideration is likely wildly out of scope for this PR, but I think it helps framing the question better.